build_tools/cmake/iree_c_module.cmake

```
@@ -0,0 +1,100 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+include(CMakeParseArguments)
+
+# iree_c_module()
+#
+# Parameters:
+# NAME: Name of target (see Note).
+# SRC: MLIR source file to compile into a c module.
+# H_FILE_OUTPUT: The H header file to output.
+# TRANSLATE_TOOL: Translation tool to invoke (CMake target).
+# FLAGS: Flags to pass to the translation tool (list of strings).
+# TESTONLY: When added, this target will only be built if user passes
+#    -DIREE_BUILD_TESTS=ON to CMake.
+#
+# Note:
+# By default, iree_c_module will create a library named ${NAME},
+# and alias target iree::${NAME}. The iree:: form should always be used.
+# This is to reduce namespace pollution.
+function(iree_c_module)
+  cmake_parse_arguments(
+    _RULE
+    "TESTONLY"
+    "NAME;SRC;H_FILE_OUTPUT;TRANSLATE_TOOL"
+    "FLAGS"
+    ${ARGN}
+  )
+
+  if(_RULE_TESTONLY AND NOT IREE_BUILD_TESTS)
+    return()
+  endif()
+
+  # Replace dependencies passed by ::name with iree::package::name
+  iree_package_ns(_PACKAGE_NS)
+  list(TRANSFORM _RULE_DEPS REPLACE "^::" "${_PACKAGE_NS}::")
+
+  # Prefix the library with the package name, so we get: iree_package_name.
+  iree_package_name(_PACKAGE_NAME)
+  set(_NAME "${_PACKAGE_NAME}_${_RULE_NAME}")
+
+  # Set defaults for FLAGS and TRANSLATE_TOOL
+  if(DEFINED _RULE_FLAGS)
+    set(_FLAGS ${_RULE_FLAGS})
+  else()
+    set(_FLAGS "-iree-vm-ir-to-c-module")
+  endif()
+  if(DEFINED _RULE_TRANSLATE_TOOL)
+    set(_TRANSLATE_TOOL ${_RULE_TRANSLATE_TOOL})
+  else()
+    set(_TRANSLATE_TOOL "iree-translate")
+  endif()
+
+  iree_get_executable_path(_TRANSLATE_TOOL_EXECUTABLE ${_TRANSLATE_TOOL})
+
+  set(_ARGS "${_FLAGS}")
+  list(APPEND _ARGS "${CMAKE_CURRENT_SOURCE_DIR}/${_RULE_SRC}")
+  list(APPEND _ARGS "-o")
+  list(APPEND _ARGS "${_RULE_H_FILE_OUTPUT}")
+
+  add_custom_command(
+    OUTPUT "${_RULE_H_FILE_OUTPUT}"
+    COMMAND ${_TRANSLATE_TOOL_EXECUTABLE} ${_ARGS}
+    # Changes to either the translation tool or the input source should
+    # trigger rebuilding.
+    DEPENDS ${_TRANSLATE_TOOL_EXECUTABLE} ${_RULE_SRC}
+  )
```

Seems like it would be just as clear to include the command args in the call to `add_custom_command`

```suggestion
  add_custom_command(
    OUTPUT "${_RULE_H_FILE_OUTPUT}"
    COMMAND 
    	${_TRANSLATE_TOOL_EXECUTABLE}
    	${_FLAGS}
    	"${CMAKE_CURRENT_SOURCE_DIR}/${_RULE_SRC}"
    	"-o" "${_RULE_H_FILE_OUTPUT}"
    # Changes to either the translation tool or the input source should
    # trigger rebuilding.
    DEPENDS ${_TRANSLATE_TOOL_EXECUTABLE} ${_RULE_SRC}
  )
```

---

This is nearly a takeover from https://github.com/google/iree/blob/881de81238971869d6d62cabf8dc0b54a17be704/build_tools/cmake/iree_bytecode_module.cmake#L64%C3%9FL77.
I have no preference, but I think it might be nice to keep it consistent?

---

Fair, though I don't think consistency in this case really buys us much. I don't care too much either way.